astropy/io/fits/hdu/hdulist.py

```diff
@@ -194,25 +273,30 @@ def __getitem__(self, key):
             hdus = super(HDUList, self).__getitem__(key)
             return HDUList(hdus)
 
-        idx = self.index_of(key)
-        return super(HDUList, self).__getitem__(idx)
+        # Originally this used recursion, but hypothetically an HDU with
+        # a very large number of HDUs could blow the stack, so use a loop
+        # instead
+        return self._try_while_unread_hdus(super(HDUList, self).__getitem__,
```

Comment:
Ah but ``__len__`` is overriden so that it reads in all the HDUs when calling len(), so that might explain it.